library/bignum.c

```diff
@@ -2391,6 +2391,16 @@ int mbedtls_mpi_gcd( mbedtls_mpi *G, const mbedtls_mpi *A, const mbedtls_mpi *B
     lz = mbedtls_mpi_lsb( &TA );
     lzt = mbedtls_mpi_lsb( &TB );
 
+    /* The loop below gives the correct result when A==0 but not when B==0.
```

Comment:
The code claims to implement [HAC 14.54](http://cacr.uwaterloo.ca/hac/about/chap14.pdf) so I don't think we need to document the correctness of what HAC tells us. HAC has a precondition that A and B are >0, so we do need to justify the correctness of the behavior in that case.

Reply:
I think that's holding this function to a much higher standard than the rest of the library, and out of scope in a security fix. But ok, done.

Reply:
I think there has been a misunderstanding. I wasn't trying to ask for a full correctness proof in the comments. I agree that it would mean a much higher standard than we require from the rest of the library. I don't even think that such verbosity is helpful.

When I was writing my first comment I was thinking about giving hints about how our code works and wrote my suggestion from that perspective.

When writing my second comment above, I already accepted your approach of documenting it relative to HAC, which I too think is a good idea.

At this point I merely was asking for writing something like:
> A precondition to HAC 14.54 is that A and B are positive. We achieve this by using functions operating on their absolute values and making checking if they are zero (for A in the loop header and for B explicitly). Another precondition listed in HAC 14.54 is that A>=B, which we don't test as it is not necessary. (This precondition is not required on line 1 or 2 of the algorithm and doesn't hold in general after line 2.)

Would you mind reworking https://github.com/ARMmbed/mbedtls/pull/4644/commits/fb0c048cb0e9c330dd93ea1f657373aef584982 in this spirit?

Also, I agree that a full correctness proof is an overkill, I think the code is much more readable without it. Could you please remove the corresponding commit: https://github.com/ARMmbed/mbedtls/pull/4644/commits/7178785b2ff1040bc2874a508ebd48c3f98b9177.?